### PR TITLE
[WIP] The Missing Ingredient in Zero-Shot Neural Machine Translation

### PR DESCRIPTION
**Pull request by francoishernandez**
This PR intends to add an implementation of the cosine similarity alignment loss introduced as a regularization term in [_The Missing Ingredient in Zero-Shot Neural Machine Translation_](https://arxiv.org/abs/1903.07091).

### CODE REVIEW

File: onmt/utils/loss.py [hunk 1]
```diff
@@ -305,6 +334,16 @@ def _compute_coverage_loss(self, std_attn, coverage_attn):
         covloss *= self.lambda_coverage
         return covloss
 
+    def _compute_cosine_loss(self, enc_src, enc_tgt):
+        max_src = enc_src.max(axis=0)[0]
+        max_tgt = enc_tgt.max(axis=0)[0]
+        cosine_loss = torch.nn.functional.cosine_similarity(
+            max_src.float(), max_tgt.float(), dim=1)
+        ones = torch.ones(cosine_loss.size()).to(cosine_loss.device)
+        cosine_loss = ones - cosine_loss
```
Comment (pltrdy) [inline onmt/utils/loss.py]: use broadcasting ? i.e. `cosine_loss = 1 - cosine_loss`

File: onmt/trainer.py [hunk 2]
```diff
@@ -70,7 +70,8 @@ def build_trainer(opt, device_id, model, fields, optim, model_saver=None):
                            model_dtype=opt.model_dtype,
                            earlystopper=earlystopper,
                            dropout=dropout,
-                           dropout_steps=dropout_steps)
+                           dropout_steps=dropout_steps,
+                           encode_tgt=True if opt.lambda_cosine > 0 else False)
```
Comment (pltrdy) [inline onmt/trainer.py]: are you sure it wont raise an issue for current checkpoint where `opt.lambda_cosine` does not exist?!

Comment (francoishernandez) [reply onmt/trainer.py]: Should be handled here:
https://github.com/OpenNMT/OpenNMT-py/blob/5ea9cb2fa0d12e1f8b3b835ecf51a4b949fc0a14/onmt/utils/parse.py#L84-L90

**Comment by francoishernandez** [comment]
*Note:*
Impact on speed is quite significant: as we need to reduce batches to make place in memory for the additional representations, we can loose up to 20-25% in training speed.
This was tested on FP32 GPUs, impact on FP16 still needs to be evaluated.

**Comment by vince62s** [comment]
For the record, we discussed offline if this should be in the code of NMTModel or Trainer.
For performance reason it needs to be in NMTModel (encoding through forward of src and tgt), but it makes the API a little less "clear". We opted for performance, but we kept the API intact when this new loss is not used.

File: onmt/trainer.py [hunk 3]
```diff
@@ -314,11 +317,21 @@ def validate(self, valid_iter, moving_average=None):
...
                 # F-prop through the model.
-                outputs, attns = valid_model(src, tgt, src_lengths,
-                                             with_align=self.with_align)
+                if self.encode_tgt:
+                    outputs, attns, enc_src, enc_tgt = valid_model(
+                        src, tgt, src_lengths,
+                        with_align=self.with_align,
+                        encode_tgt=self.encode_tgt)
+                else:
+                    output, attns = valid_model(
```
Comment (vince62s) [inline onmt/trainer.py]: là aussi outputs ?

Comment (francoishernandez) [reply onmt/trainer.py]: ...

File: onmt/models/model.py [hunk 4]
```diff
@@ -17,7 +18,8 @@ def __init__(self, encoder, decoder):
         self.encoder = encoder
         self.decoder = decoder
 
-    def forward(self, src, tgt, lengths, bptt=False, with_align=False):
+    def forward(self, src, tgt, lengths, bptt=False,
+                with_align=False, encode_tgt=False):
```
Comment (vince62s) [inline onmt/models/model.py]: @francoishernandez second thought.... this changes also the API, right ?
